decomposition_tensor.py

```diff
@@ -0,0 +1,162 @@
+#
+# What is progressive lowering tensor and how does this compare?
+# - TODO
+
+aten = torch.ops.aten
+
+skip_list = [aten.add.Tensor, aten.to.dtype, aten.div.Tensor, aten.clamp.default]
+
+# 1) Wrapper subclass inline version
+class DecompositionTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, e):
+        r = torch.Tensor._make_wrapper_subclass(cls, e.shape, dtype=e.dtype, requires_grad=False)
+        r.elem = e
+        return r
+
+    # We may be able to remove this line in the future when Ed's PR lands
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore
+        def unwrap(e):
+            return e.elem if isinstance(e, DecompositionTensor) else e
+
+        def wrap(e):
+            return DecompositionTensor(e) if isinstance(e, torch.Tensor) else e
+
+        if func in skip_list:
+            # Check skip-list first, so "backend" has a chance to see non-decomposed ops
+            return tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))
+        elif func in decomposition_table:
+            return decomposition_table[func](*args, **kwargs)
+        else:
+            raise NotImplementedError(f"{func.__name__} does not have a decomposition and is not in skip_list")
+
+# 2) Decorator (Alban's suggestion)
+def decompose(skip_list, missing_ops=None):
+    def _decompose(f):
+        @wraps(f)
+        def wrapper(cls, func, types, args=(), kwargs=None):
+            if func in skip_list:
+                # Functions that the layers below are able to handle
+                return f(cls, func, types, args, kwargs)
+            elif func in decomposition_table:
+                return decomposition_table[func](*args, **kwargs)
+            else:
+                if missing_ops is not None:
+                    missing_ops.add(func.__name__)
+                    return f(cls, func, types, args, kwargs)
+                else:
+                    raise NotImplementedError(f"{func.__name__} does not have a decomposition and is not in skip_list")
+        return wrapper
+    return _decompose
+
+# 2.1) Using the decorator
+class DecompositionTensor2(torch.Tensor):
+    @staticmethod
+    def __new__(cls, e):
+        r = torch.Tensor._make_wrapper_subclass(cls, e.shape, dtype=e.dtype, requires_grad=False)
+        r.elem = e
+        return r
+
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+    @classmethod
+    @decompose(skip_list, missing_ops=None)
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore
+        def unwrap(e):
+            return e.elem if isinstance(e, DecompositionTensor2) else e
+
+        def wrap(e):
+            return DecompositionTensor2(e) if isinstance(e, torch.Tensor) else e
+
+        return tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))
+
+# 3) Version using inheritance
```

Comment:
I'm generally against implementing this kind of extra functionality with inheritance. Better to make sure there is some sort of subtyping relation if you're going to use inheritance.